### PR TITLE
New logger with color and specific types. Code cleanup 

### PR DESCRIPTION
**Pull request by IronMine**
In this PR there is : 
- Removing some unused functions
- A new logger system with colors
- A Warning specific logger ( yellow )
- An Eroor specific logger ( red ) 
- Fixing broken CanyonModded link url to jenkins
- Optimizing some curl calls
- A little bit of code cleanup
- Some proposal ( start-finalExec and start-setupServerProperties)

### CODE REVIEW

**Comment by itzg** [comment]
This is a lot of changes for one PR. I know this is hard after the fact, but is there any way you could break this up into a few smaller ones?

**Comment by IronMine** [comment]
will do !

**Comment by IronMine** [comment]
is there some things i listed that you do not want ?

**Comment by itzg** [comment]
Glancing at the changes I like all of it 😃 

Actually, keep the PR as is and I'll review more closely when I get a span of time. With that I can raise any specific concerns I find, if any.

**Comment by IronMine** [comment]
Nice perfect ! 
As you want it is a huge PR i do understand if you would prefer it separate in multiple small ones

**Comment by itzg** [comment]
With new formatted logging, the output from `DEBUG` set to "true" is getting noisier, such as:
```
[init] 2024-10-20 16:46:51+00:00
+ logError '         SPONGEVANILLA, CUSTOM, MAGMA, MOHIST, CATSERVER, AIRPLANE, PUFFERFISH,'
++ getErrorColoredLogString '         SPONGEVANILLA, CUSTOM, MAGMA, MOHIST, CATSERVER, AIRPLANE, PUFFERFISH,'
++ echo '\033[0;31m[ERROR]          SPONGEVANILLA, CUSTOM, MAGMA, MOHIST, CATSERVER, AIRPLANE, PUFFERFISH, \033[0m'
+ log
+ local oldState
++ shopt -po xtrace
+ oldState='set -o xtrace'
+ shopt -u -o xtrace
```

File: scripts/start-utils [hunk 1]
```diff
@@ -109,10 +117,86 @@ function log() {
...
+}
+
+function error() {
+  echo -e $(getErrorColoredLogString "$*")
+}
+
+function logError() {
+  log $(getErrorColoredLogString "$*")
+}
+
```
Comment (itzg) [inline scripts/start-utils:150]: Why both `error` and `logError`? I'd say go with `logError` and just collapse the color coding into it.

Same for warning.

File: bin/mc-send-to-console [hunk 2]
```diff
@@ -3,26 +3,24 @@
...
   exit 1
 fi
 
 if [ ! -p "${CONSOLE_IN_NAMED_PIPE}" ]; then
-  echo "ERROR: named pipe ${CONSOLE_IN_NAMED_PIPE} is missing"
+  error "Named pipe ${CONSOLE_IN_NAMED_PIPE} is missing"
   exit 1
 fi
 
+commandSuffix="${CONSOLE_IN_NAMED_PIPE:-/tmp/minecraft-console-in}"
```
Comment (itzg) [inline bin/mc-send-to-console:20]: CONSOLE_IN_NAMED_PIPE is already a variable initialized on line 3. No need to assign again to a variable.

File: files/auto/stop.sh [hunk 3]
```diff
@@ -6,8 +6,9 @@ if isTrue "${DEBUG_AUTOSTOP}"; then
 fi
 
 logAutostopAction "Stopping Java process"
+commandToExecute="pkill -f --signal SIGTERM mc-server-runner"
 if isTrue "${AUTOSTOP_PKILL_USE_SUDO:-false}"; then
-  sudo pkill -f --signal SIGTERM mc-server-runner
-else
-  pkill -f --signal SIGTERM mc-server-runner
+  commandToExecute="sudo $commandToExecute"
```
Comment (itzg) [inline files/auto/stop.sh:11]: The restructuring loses clarity of if-else

```suggestion
  sudo $commandToExecute
else
  $commandToExecute
```

File: bin/mcstatus [hunk 4]
```diff
@@ -11,8 +11,9 @@ addr="$1"
...
 read -a parts <<< "${addr}"
+commandToExecute="exec mc-monitor status --host ${parts[0]}"
 if [[ ${#parts[*]} -gt 1 ]]; then
-  exec mc-monitor status --host ${parts[0]} --port ${parts[1]}
-else
-  exec mc-monitor status --host ${parts[0]}
+    commandToExecute="${commandToExecute} --port ${parts[1]}"
 fi
+
+$commandToExecute
```
Comment (itzg) [inline bin/mcstatus:19]: Use `args` approach like I do elsewhere.

File: build/alpine/install-packages.sh
Comment (itzg) [inline build/alpine/install-packages.sh:1]: Awkward refactoring. Revert this file.

File: build/ol/install-packages.sh
Comment (itzg) [inline build/ol/install-packages.sh:1]: Awkward refactoring. Revert this file.

File: build/run.sh [hunk 5]
```diff
@@ -4,4 +4,13 @@ set -e
...
+"$(dirname "$0")/${distro}/$1".sh
+
+if [ "$1" = "install-packages" ]; then
+# Set Git credentials globally
+    cat <<EOF >> /etc/gitconfig
+[user]
+	name = Minecraft Server on Docker
+	email = server@example.com
+EOF
+fi
```
Comment (itzg) [inline build/run.sh:16]: This is awkward. Original was nice and opaque to the sub-command called.

File: build/ubuntu/install-packages.sh
Comment (itzg) [inline build/ubuntu/install-packages.sh:1]: same

File: scripts/start-autopause
Comment (itzg) [inline scripts/start-autopause:1]: Too many changes. Just drop it back to logging call changes.

File: scripts/start-utils [hunk 6]
```diff
@@ -109,10 +117,86 @@ function log() {
...
+function checkIfNumeric() {
+  [[ $1 =~ ^[0-9]+$ ]]
+}
+
+function checkIfNumericElseSetToDefault() {
+  local var_name="$1"
+  local default_value="$2"
+  
+  if ! checkIfNumeric ${!var_name} ; then
+    eval "$var_name=$default_value"
```
Comment (itzg) [inline scripts/start-utils:184]: Are you absolutely sure this works?

File: scripts/start-utils [hunk 7]
```diff
@@ -109,10 +117,86 @@ function log() {
...
+    export "$var_name"
+    logAutoPauseAndStopChange "$var_name is not numeric" $default_value
```
Comment (itzg) [inline scripts/start-utils:186]: Shouldn't reference "auto pause" within this generic start-utils. Use logWarning

File: scripts/start-utils [hunk 8]
```diff
@@ -109,10 +117,86 @@ function log() {
...
+  fi
+}
+
+function checkIfNotZeroElseSetToDefault() {
+  local var_name="$1"
+  local default_value="$2"
+  if [ "${!var_name}" -eq "0" ] ; then
+    eval "$var_name=$default_value"
+    export "$var_name"
+    logAutoPauseAndStopChange "$var_name must not be 0" $default_value
```
Comment (itzg) [inline scripts/start-utils:196]: same

File: scripts/start-utils [hunk 9]
```diff
@@ -18,6 +18,14 @@ function get_from_gh() {
   fi
 }
 
+function grabServerAndExport() {
```
Comment (itzg) [inline scripts/start-utils:21]: Call this something clearer like `applyResultsFile`

File: scripts/start-configuration [hunk 10]
```diff
@@ -90,13 +90,15 @@ fi
 ##########################################
 # Auto-pause/stop
 
+INCOMPATIBLE_LOG="is incompatible with"
```
Comment (itzg) [inline scripts/start-configuration:93]: Too much refactoring

File: scripts/start-configuration [hunk 11]
```diff
@@ -166,118 +168,122 @@ case "${TYPE^^}" in
...
+      MOD_PLATFORM_SHORT="Modrinth"
       ;;
 
     *)
-      log "ERROR; Invalid MOD_PLATFORM: '$MOD_PLATFORM'"
+      logError "Invalid MOD_PLATFORM: '$MOD_PLATFORM'"
       exit 1
       ;;
   esac
+  exec "${SCRIPTS:-/}start-deploy${MOD_PLATFORM_SHORT}" "$@"
```
Comment (itzg) [inline scripts/start-configuration:194]: Too much refactoring

File: scripts/start-configuration [hunk 12]
```diff
@@ -166,118 +168,122 @@ case "${TYPE^^}" in
...
 fi
 
 log "Resolving type given ${TYPE}"
+TYPE_SCRIPT_SUFFIX=""
```
Comment (itzg) [inline scripts/start-configuration:198]: Too much refactoring

File: scripts/start-deployAutoCF [hunk 13]
```diff
@@ -57,15 +57,11 @@ setArg --exclude-include-file CF_EXCLUDE_INCLUDE_FILE
 setArg --downloads-repo CF_DOWNLOADS_REPO
 
 if ! mc-image-helper install-curseforge "${args[@]}"; then
-    log "ERROR failed to auto-install CurseForge modpack"
+    logErrorFailedTo "auto-install CurseForge modpack"
```
Comment (itzg) [inline scripts/start-deployAutoCF:60]: use logError

File: scripts/start-deployBukkitSpigot [hunk 14]
```diff
@@ -6,47 +6,56 @@ isDebugging && set -x
 
 set -eo pipefail
 
+SPIGOT_BUILD_LOG="/data/spigot_build.log"
+
+function failedSpigotBuild {
```
Comment (itzg) [inline scripts/start-deployBukkitSpigot:11]:
```suggestion
function handleFailedSpigotBuild {
```

File: scripts/start-deployBukkitSpigot [hunk 15]
```diff
@@ -6,47 +6,56 @@ isDebugging && set -x
...
+  logErrorFailedTo "build Spigot"
+  cat ${SPIGOT_BUILD_LOG}
+  exit 1
+}
+
+
 function buildSpigotFromSource {
+  local TEMP_DIR="/data/temp"
```
Comment (itzg) [inline scripts/start-deployBukkitSpigot:19]: Lower camel case for locals. External vars are upper snake case.
```suggestion
  local tempDir="/data/temp"
```

File: scripts/start-deployBukkitSpigot [hunk 16]
```diff
@@ -6,47 +6,56 @@ isDebugging && set -x
 
 set -eo pipefail
 
+SPIGOT_BUILD_LOG="/data/spigot_build.log"
```
Comment (itzg) [inline scripts/start-deployBukkitSpigot:9]:
```suggestion
spigotBuildLog="/data/spigot_build.log"
```

File: scripts/start-deployCF [hunk 17]
```diff
@@ -60,12 +60,13 @@ downloadModpack() {
       # If not an absolute path, assume file is in "/data"
       srv_modpack=/data/${srv_modpack}
   fi
+  FTBServerModpackLog="FTB server modpack: ${srv_modpack}"
   if [[ ! -f "${srv_modpack}" ]]; then
-      log "FTB server modpack ${srv_modpack} not found."
+      log "$FTBServerModpackLog not found."
```
Comment (itzg) [inline scripts/start-deployCF:65]: logError

File: scripts/start-deployCF [hunk 18]
```diff
@@ -60,12 +60,13 @@ downloadModpack() {
...
       exit 2
   fi
   if [[ ! "${srv_modpack: -4}" == ".zip" ]]; then
-      log "FTB server modpack ${srv_modpack} is not a zip archive."
+      log "$FTBServerModpackLog is not a zip archive."
```
Comment (itzg) [inline scripts/start-deployCF:69]: logError

File: scripts/start-deployCanyon [hunk 19]
```diff
@@ -9,20 +9,19 @@ isDebugging && set -x
 : "${CANYON_BUILD:=lastSuccessfulBuild}"
 
 if [ "${VERSION}" != "b1.7.3" ]; then
-  log "ERROR: Canyon server type only supports VERSION=b1.7.3"
+  logError "Canyon server type only supports VERSION=b1.7.3"
   exit 1
 fi
 
-canyonJob="https://jenkins.glass-launcher.net/job/Canyon"
+canyonJob="https://canyonmodded.p0rtal.gay/job/Canyonmodded"
```
Comment (itzg) [inline scripts/start-deployCanyon:16]: Revert URL change

File: scripts/start-deployFTBA [hunk 20]
```diff
@@ -9,42 +9,51 @@ ftbInstallMarker=".ftb-installed"
 isDebugging && set -x
 set -e
 
+function logFTBInstaller(){
+   log "Downloading $* installer"
+}
```
Comment (itzg) [inline scripts/start-deployFTBA:14]: Too much refactoring

File: scripts/start-deployFTBA [hunk 21]
```diff
@@ -63,16 +72,17 @@ else
...
 export VERSION
 
+baseFileName="forge-${mcVersion}-${forgeVersion}"
 variants=(
-  "forge-${mcVersion}-${forgeVersion}.jar"
-  "forge-${mcVersion}-${forgeVersion}-universal.jar"
-  "forge-${mcVersion}-${forgeVersion}-${mcVersion}-universal.jar"
+  "${baseFileName}.jar"
+  "${baseFileName}-universal.jar"
+  "${baseFileName}-${mcVersion}-universal.jar"
```
Comment (itzg) [inline scripts/start-deployFTBA:85]: Too much refactoring

File: scripts/start-deployFabric [hunk 22]
```diff
@@ -12,44 +12,42 @@ export TYPE=FABRIC
 
 resultsFile=/data/.install-fabric.env
 
+function mc-image-helper-fabric() {
+  mc-image-helper install-fabric-loader \
+    --results-file=${resultsFile} \
+    "$@"
+}
+
```
Comment (itzg) [inline scripts/start-deployFabric:20]: Too much refactoring

File: scripts/start-deployCanyon (hunk 19 above)
Comment (IronMine) [reply scripts/start-deployCanyon:16]: it does not work anymore with this url but ok

File: scripts/start-utils (hunk 1 above)
Comment (IronMine) [reply scripts/start-utils:150]: it's for those file : /bin/mc-send-to-console
So they can use the new log system without using the log function that add [init] before all logged lines

File: scripts/start-autopause
Comment (IronMine) [reply scripts/start-autopause:1]: so i revert ?

File: scripts/start-utils (hunk 6 above)
Comment (IronMine) [reply scripts/start-utils:184]: yes tested it with multiple possibilities: 
`. "scripts/start-utils"

export TEST_VALUE="666666"

isNumericElseSetToDefault TEST_VALUE 10
checkIfNotZeroElseSetToDefault TEST_VALUE 10

export TEST_VALUE="êrer"
isNumericElseSetToDefault TEST_VALUE 10
checkIfNotZeroElseSetToDefault TEST_VALUE 10

export TEST_VALUE="6708907e"
isNumericElseSetToDefault TEST_VALUE 10
checkIfNotZeroElseSetToDefault TEST_VALUE 10

export TEST_VALUE="0"
isNumericElseSetToDefault TEST_VALUE 10
checkIfNotZeroElseSetToDefault TEST_VALUE 10

export TEST_VALUE="-1"
isNumericElseSetToDefault TEST_VALUE 10
checkIfNotZeroElseSetToDefault TEST_VALUE 10

export TEST_VALUE=9
isNumericElseSetToDefault TEST_VALUE 10
checkIfNotZeroElseSetToDefault TEST_VALUE 10
`

File: scripts/start-utils (hunk 1 above)
Comment (itzg) [reply scripts/start-utils:150]: Thanks, I forgot that the logs get prefixed with `[init]`. In that case, they're fine as is.

File: scripts/start-autopause
Comment (itzg) [reply scripts/start-autopause:1]: Yes, but the `error` and `warn` changes are fine.

File: bin/mcstatus (hunk 4 above)
Comment (IronMine) [reply bin/mcstatus:19]: ohh okay i see thanks !

File: scripts/start-deployForge [hunk 23]
```diff
@@ -6,36 +6,29 @@
 # shellcheck source=start-utils
 . "${SCRIPTS:-$(dirname "$0")}/start-utils"
 isDebugging && set -x
+resultsFile=/data/.run-forge.env
 
-if [[ ${FORGE_INSTALLER} ]]; then
-  if ! mc-image-helper install-forge \
+function mc-image-helper-forge() {
```
Comment (itzg) [inline scripts/start-deployForge:11]: This is a nice refactoring pattern

File: scripts/start-utils [hunk 24]
```diff
@@ -109,10 +117,78 @@ function log() {
...
+  echo $command
+}
+
+function getErrorColoredLogString() {
+  echo "\033[0;31mERROR $* \033[0m"
+}
+
+function getWarningColoredLogString() {
+  echo "\033[0;33mWARN $* \033[0m"
+}
```
Comment (itzg) [inline scripts/start-utils:141]: To ensure maximum portability:

```suggestion
# Refer to https://unix.stackexchange.com/a/10065/102376
function isTerminal() {
  if test -t 1; then
    # see if it supports colors...
    ncolors=$(tput colors)
    test -n "$ncolors" && test "$ncolors" -ge 8
  else
    return 1
  fi
}

if isTerminal; then
  normal="$(tput sgr0)"
  red="$(tput setaf 1)"
  yellow="$(tput setaf 3)"

  function getErrorColoredLogString() {
    echo "${red}ERROR $*${normal}"
  }

  function getWarningColoredLogString() {
    echo "${yellow}WARN $*${normal}"
  }
else
  function getErrorColoredLogString() {
    echo "ERROR $*"
  }

  function getWarningColoredLogString() {
    echo "WARN $*"
  }
fi
```

File: scripts/start-utils [hunk 25]
```diff
@@ -109,10 +117,78 @@ function log() {
...
+
+function error() {
+  echo -e $(getErrorColoredLogString "$*")
+}
+
+function logError() {
+  if isDebugging; then
+    set -x
+    return 0
+  fi
```
Comment (itzg) [inline scripts/start-utils:151]: Now it won't log at all other than the debugging output, right?

File: scripts/start-utils [hunk 26]
```diff
@@ -109,10 +117,78 @@ function log() {
...
+  log $(getErrorColoredLogString "$*")
+}
+
+function warning() {
+  if isDebugging; then
+    set -x
+    return 0
+  fi
```
Comment (itzg) [inline scripts/start-utils:159]: same?

File: scripts/start-setupMounts [hunk 27]
```diff
@@ -20,53 +20,44 @@ if isTrue "${SYNC_SKIP_NEWER_IN_DESTINATION}"; then
...
 fi
 
+subcommand=sync
+
 if isTrue "${REPLACE_ENV_DURING_SYNC}"; then
-  subcommand=sync-and-interpolate
-else
-  subcommand=sync
+  subcommand+=-and-interpolate
 fi
```
Comment (itzg) [inline scripts/start-setupMounts:27]: Please revert to improve readability

File: scripts/start-deployNeoForge [hunk 28]
```diff
@@ -15,15 +15,11 @@ if ! mc-image-helper install-neoforge \
   --minecraft-version="${VERSION}" \
   --neoforge-version="${NEOFORGE_VERSION}" \
   --force-reinstall="${NEOFORGE_FORCE_REINSTALL}"; then
-    log "ERROR failed to install Forge"
+    logError "Failed to installForge"
```
Comment (itzg) [inline scripts/start-deployNeoForge:18]:
```suggestion
    logError "Failed to install NeoForge"
```

File: scripts/start-finalExec [hunk 29]
```diff
@@ -77,7 +81,7 @@ elif isFamily VANILLA && versionLessThan 1.17; then
   patchLog4jConfig log4j2_112-116.xml https://launcher.mojang.com/v1/objects/02937d122c86ce73319ef9975b58896fc1b491d1/log4j2_112-116.xml
 # See https://purpurmc.org/docs/Log4j/
 elif isType PURPUR && versionLessThan 1.17; then
-  patchLog4jConfig purpur_log4j2_1141-1165.xml https://purpurmc.org/docs/xml/purpur_log4j2_1141-1165.xml
+  patchLog4jConfig purpur_log4j2_1141-1165.xml https://purpssurmc.org/docs/xml/purpur_log4j2_1141-1165.xml
```
Comment (itzg) [inline scripts/start-finalExec:84]: This URL didn't work for me

File: scripts/start-deployCanyon [hunk 30]
```diff
@@ -9,20 +9,19 @@ isDebugging && set -x
 : "${CANYON_BUILD:=lastSuccessfulBuild}"
 
 if [ "${VERSION}" != "b1.7.3" ]; then
-  log "ERROR: Canyon server type only supports VERSION=b1.7.3"
+  logError "Canyon server type only supports VERSION=b1.7.3"
   exit 1
 fi
 
 canyonJob="https://jenkins.glass-launcher.net/job/Canyon"
```
[the remark this answers is not in the document]
Comment (IronMine) [reply scripts/start-deployCanyon:16]: i mentionned it in the commit but i am not realy smart about that :)

File: scripts/start-deployForge (hunk 23 above)
Comment (IronMine) [reply scripts/start-deployForge:11]: Thanks

File: scripts/start-finalExec (hunk 29 above)
Comment (IronMine) [reply scripts/start-finalExec:84]: this is a mistake from me sorry !

File: scripts/start-utils (hunk 24 above)
Comment (itzg) [reply scripts/start-utils]: Yep, very good

**Review by itzg — APPROVED** [review]
Thanks so much for working with me on all the feedback I provided.

**Review by itzg — COMMENTED** [review]
I was doing some final testing and noticed now that the logged string just disappears for errors (and probably warnings):

```
docker run --rm -it itzg/minecraft-server
[init] Running as uid=1000 gid=1000 with /data as 'drwxr-x--- 2 1000 1000 4096 Oct 22 01:17 /data'
[init] 
[init] 
[init] 
[init] 
[init] 
[init] 
```

**Comment by itzg** [comment]
Ah, IntelliJ was actually reporting the problem:

image

**Comment by IronMine** [comment]
> Thanks so much for working with me on all the feedback I provided.

your are welcome and i apologize for all the time you loose on my PR's ^^

**Comment by IronMine** [comment]
like that ?

File: scripts/start-utils [hunk 31]
```diff
@@ -109,10 +117,107 @@ function log() {
...
+  function getErrorColoredLogString() {
+    echo "$errorLogTag $*"
+  }
+  function getWarningColoredLogString() {
+    echo "$warningLogTag $*"
+  }
+fi
+
+function error() {
+  echo -e $(getErrorColoredLogString "$*")
```
Comment (itzg) [inline scripts/start-utils:169]:
```suggestion
  echo -e "$(getErrorColoredLogString "$*")"
```

File: scripts/start-utils [hunk 32]
```diff
@@ -109,10 +117,107 @@ function log() {
...
+  if isDebugging; then
+    set -x
+  fi
+}
+
+function warning() {
+  if isDebugging; then
+    set +x
+  fi
+  echo -e $(getWarningColoredLogString "$*")
```
Comment (itzg) [inline scripts/start-utils:186]:
```suggestion
  echo -e "$(getWarningColoredLogString "$*")"
```

File: scripts/start-utils [hunk 33]
```diff
@@ -109,10 +117,107 @@ function log() {
...
+  [[ $1 =~ ^[0-9]+$ ]]
+}
+
+function isNumericElseSetToDefault() {
+  local var_name="$1"
+  local default_value="$2"
+  
+  if ! isNumeric ${!var_name} ; then
+    eval "$var_name=$default_value"
+    export "$var_name"
```
Comment (itzg) [inline scripts/start-utils:206]: ShellCheck is reporting a warning here:

image

Comment (IronMine) [reply scripts/start-utils:206]: yes because i am doing the thing it thinks i do not want to do

**Review by itzg — APPROVED** [review]

**Pull request merged**